src/main/scala/monocly/Optic.scala

```diff
@@ -1,65 +1,205 @@
+    asInstanceOf[Optic[ThisCan, S1, A1]].andThen(evIndex.index(i))
 
-extension [S, T, A, B] (optic: POptic[GetOneOrMore, S, T, A, B])
-  inline def getOneOrMore: S => NonEmptyList[A] = optic.getter.foldMap1[NonEmptyList[A]](NonEmptyList(_, Nil))
+  def modify(f: A => B)(using ThisCan <:< Modify): S => T = 
+    impl.modify(f)
 
-extension [S, T, A, B] (optic: POptic[GetOption, S, T, A, B])
-  inline def getOption: S => Option[A] = optic.getter.getOption
+  def modifyA[F[+_]: Applicative](f: A => F[B])(s: S)(using ThisCan <:< GetMany & Modify): F[T] = 
+    impl.modifyA(f)(s)
 
-extension [S, T, A, B] (optic: POptic[GetOption & Modify, S, T, A, B])
-  inline def getOrModify: S => Either[T, A] = 
-    s => optic.getter.getOption(s).fold(Left(???))(Right.apply) 
+  def nonEmptyFoldMap[M: Semigroup](f: A => M)(s: S)(using ThisCan <:< GetOneOrMore): M = 
+    impl.nonEmptyFoldMap(f)(s)
 
-extension [ThisCan, S, A] (optic: Optic[ThisCan, S, A])
-  def index[I, A1](i: I)(using evIndex: Index[A, I, A1]): Optic[ThisCan | (GetOption & Modify), S, A1] = 
-    optic.andThen(evIndex.index(i))
+  def nonEmptyModifyA[F[+_]: Apply](f: A => F[B])(s: S)(using ThisCan <:< GetOneOrMore & Modify): F[T] = 
+    impl.nonEmptyModifyA(f)(s)
 
-extension [S, T, A, B] (optic: POptic[Modify, S, T, A, B])
-  inline def modify(f: A => B): S => T = optic.setter.modify(f)
+  def replace(b: B)(using ThisCan <:< Modify): S => T = 
+    impl.modify(_ => b)
 
-extension [S, T, A, B] (optic: POptic[GetMany & Modify, S, T, A, B])
-  inline def modifyA[F[_]: Applicative](f: A => F[B])(s: S): F[T] = ???
+  def reverseGet(using ThisCan <:< ReverseGet): B => T = 
+    impl.reverseGet
 
-extension [S, T, A, B] (optic: POptic[Modify, S, T, A, B])
-  inline def replace(b: B): S => T = optic.setter.modify(_ => b)
-  
-extension [S, T, A, B] (optic: POptic[ReverseGet, S, T, A, B])
-  inline def reverseGet: B => T = optic.setter.reverseGet
+  def some[A1, B1](using A <:< Option[A1], Option[B1] <:< B): POptic[ThisCan | GetOption, S, T, A1, B1] = 
+    asInstanceOf[POptic[ThisCan, S, T, Option[A1], Option[B1]]].andThen(std.option.pSome)
+
+  def withDefault[A1 >: A, B1 <: B](defaultValue: A1)(using A <:< Option[A1]): POptic[ThisCan | (Get & ReverseGet), S, T, A1, B1] = 
+    val iso = POptic.thatCan.convertBetween[Option[A1], Option[B1], A1, B1](_.getOrElse(defaultValue))(Some.apply)
+    asInstanceOf[POptic[ThisCan, S, T, Option[A1], Option[B1]]].andThen(iso)
+
+end POptic
 
-extension [ThisCan, S, T, A, B] (optic: POptic[ThisCan, S, T, Option[A], Option[B]])
-  def some: POptic[ThisCan | GetOption, S, T, A, B] = optic.andThen(std.option.pSome)
+object POptic:
+  def id[A, B]: POptic[Get & ReverseGet, A, B, A, B] = 
+    thatCan.convertBetween[A, B, A, B](identity)(identity)
+
+  object thatCan:
+    def modify[S, T, A, B](_modify: (A => B) => S => T): POptic[Modify, S, T, A, B] = 
```

Comment:
I am not sure about `modify`, it seems to be similar to `edit`. is there a way to differentiate reaod-only, write-only and normal optics?

Reply:
I like `edit` for all the reasons you mentioned.